8304425: ClassHierarchyResolver from Reflection

liach opened the pull request:
Add API to explore Class Hierarchy with a `ClassLoader` or a `Lookup` with proper privileges, with tests.

This addition is useful in case classes at runtime are not loaded from the system class loader, such as Proxy. This is also useful to APIs that generate bytecode with a `Lookup` object, such as a custom single-abstract-method class implementations from a method handle.

See https://mail.openjdk.org/pipermail/classfile-api-dev/2023-March/000249.html as well.

Current questions, which I wish to discuss with @asotona:
1. Should the resolver fail fast on `IllegalAccessException` from the lookup? This usually indicates the hierarchy resolver is set up improperly, and proceeding may simply yield verification errors in class loading that are hard to track. For bytecode-generating APIs, throwing access errors for the Lookup eagerly is also more preferable than later silent generation failure.
2. Whether the default resolver should be reading from jrt alone, reflection alone, or jrt then reflection. I personally believe reflection alone is more reliable, for classes may redefined with instrumentation or jfr, which may not be reflected in the system resources.
3. In addition, I don't think chaining system class loader reflection after system resource retrieval is really meaningful: is there any case where reflection always works while the system resource retrieval always fails?

asotona commented:
> @asotona So should I simply throw an `IllegalAccessError` when the Lookup encounters a `IllegalAccessException`, or should I return null? I favor throwing an `IllegalAccessError` as the lookup represents bytecode accessibility, and shall it fail to access, the generated bytecode will fail to access the specified superclass as well.

Yes, I also prefer throwing `IllegalAccessError` to directly indicate what is wrong instead of later indirect exceptions. I think `IllegalAccessError` is not a case for fallback resolver, so it should not be masked.

liach commented:
After thinking, I chose `IllegalArgumentException` for an error is not indicative here (as this should not throw an `IncompatibleClassChangeError`, the superclass of `IllegalAccessError`). I will keep the delegation to system classloader for default resolver for now, though it appears this is subject to discussion on the mailing list.

asotona commented on test/jdk/jdk/classfile/ClassHierarchyInfoTest.java:
```diff
@@ -100,22 +100,27 @@ void testProvideCustomClassStreamResolver() throws Exception {
+        assertThrows(IllegalArgumentException.class, () -> transformAndVerify(ClassHierarchyResolver.ofReflection(lookup)));
     }
 
     void transformAndVerify(ClassHierarchyResolver res) throws Exception {
```
It might be useful to cache any provided resolver.

liach commented:
@asotona Just curious, what's the current state of our new API model of caching class hierarchy info in a Classfile context object as we've discussed on the mailing list? Will you create a patch, or should I update this patch to that model?

liach commented:
> > @asotona Just curious, what's the current state of our new API model of caching class hierarchy info in a Classfile context object as we've discussed on the mailing list? Will you create a patch, or should I update this patch to that model?
> 
> In the discussion I tried to fine-tune exact naming of the factory methods and the default. Mainly to differentiate when the class is parsed as a resource `ofResourceParsing(ClassLoader loader)` and when it is loaded `ofClassLoading(ClassLoader loader)`
> 
> Otherwise this patch looks good.
> 
> The other part of the discussion with proposed `ClassfileReaderWriter` model is out of the scope of this topic, it has much bigger impact on all existing code and I would deferred it at least until this and other open PRs are merged.

Thanks for your evaluation, Adam. I have updated the method names to `ofClassLoading` and `ofResourceParsing`, and incorporated the idea of SecurityManager bypassing from https://github.com/openjdk/jdk/pull/13197/commits/cc6994202c93edd0c98e6fb0bb5f0cbe1da7df8e#diff-f044f7ab894bab3b36ad9f29e0b3f541df5246c90159635159343ce20e02c12cR54-R63 to ensure system libraries can use the hierarchy resolver all fine. I've ran `test/jdk/jdk/classfile` tests on my machine. Can you review again?

asotona commented:
/reviewers 2

asotona commented:
/reviewers 2 reviewer

asotona commented on src/java.base/share/classes/jdk/internal/classfile/impl/ClassHierarchyImpl.java:
```diff
@@ -101,66 +108,90 @@ public boolean isAssignableFrom(ClassDesc thisClass, ClassDesc fromClass) {
+            public InputStream apply(ClassDesc cd) {
+                var sm = System.getSecurityManager();
+                if (sm != null) {
+                    return AccessController.doPrivileged(new PrivilegedAction<>() {
```
This is out of my competency, need another reviewer for use of `PriviledgedAction` in this context.

liach commented:
A brief summary of the latest update:
1. I've moved to implement all 3 API changes to ClassHierarchyResolver discussed on classfile-api-dev
2. I've dropped SecurityManager content; if they are needed by another patch, those patches can add on their own.

This will be in conflict with the Classfile object update; I can update accordingly if that one is integrated first.

asotona commented on src/java.base/share/classes/jdk/internal/classfile/impl/ClassHierarchyImpl.java:
```diff
@@ -101,88 +113,130 @@ public boolean isAssignableFrom(ClassDesc thisClass, ClassDesc fromClass) {
 
+        @Override
+        public ClassHierarchyInfo getClassInfo(ClassDesc classDesc) {
+            var ret = resolvedCache.computeIfAbsent(classDesc, new Function<>() {
```
couldn't this function be cached?

asotona commented on test/jdk/jdk/classfile/ClassHierarchyInfoTest.java:
```diff
@@ -25,16 +25,30 @@
+ *          java.base/jdk.internal.classfile.instruction
+ *          java.base/jdk.internal.classfile.impl
+ *          java.base/jdk.internal.classfile.impl.verifier
+ *          java.base/jdk.internal.classfile.java.lang.constant
```
jdk.internal.classfile.java.lang.constant package has been removed

asotona reviewed (changes requested):
Globas static cache needs to be synchronised If this PR goes in before #14180 
Or a `defaultResolver()` should create a new cache instance (and document it is thread-unsafe) if it goes after #14180

liach commented:
Done. I've synchronized the global static cache for now, and thanks for the review!

liach commented:
Integrating this before #14180. The test failure appears to be unrelated from ScopedValues, which is not related to this patch.

/integrate

asotona commented:
/sponsor

ExE-Boss commented on src/java.base/share/classes/jdk/internal/classfile/impl/ClassHierarchyImpl.java:
```diff
@@ -101,88 +120,134 @@ public boolean isAssignableFrom(ClassDesc thisClass, ClassDesc fromClass) {
     public static final class CachedClassHierarchyResolver implements ClassHierarchyResolver {
-
-        //this instance should never appear in the cache nor leak out
+        //this instance should leak out, appears only in cache in order to utilize Map.computeIfAbsent
```
Missing “not”:
```suggestion
        // this instance should not leak out, appears only in cache in order to utilize Map.computeIfAbsent
```

ExE-Boss commented on src/java.base/share/classes/jdk/internal/classfile/impl/Util.java:
```diff
@@ -84,9 +84,26 @@ public static int maxLocals(int flags, MethodTypeDesc mDesc) {
+     * @return the fully qualified binary name
+     */
+    public static String toBinaryName(String descOrInternalName) {
+        if (descOrInternalName.startsWith("L")) {
+            // descriptors of classes or interfaces
+            if (descOrInternalName.length() <= 2 || !descOrInternalName.endsWith(";")) {
+                throw new IllegalArgumentException(descOrInternalName);
+            }
+            return descOrInternalName.substring(1, descOrInternalName.length() - 1).replace('/', '.');
+        } else {
```
Note that this won’t work right if this method is called with the internal name of a class whose FQCN starts with `L`, the correct implementation would be:
```suggestion
        if (descOrInternalName.startsWith("L") && descOrInternalName.endsWith(";")) {
            // descriptors of classes or interfaces
            if (descOrInternalName.length() <= 2) {
                throw new IllegalArgumentException(descOrInternalName);
            }
            return descOrInternalName.substring(1, descOrInternalName.length() - 1).replace('/', '.');
        } else {
```

liach replied:
Honestly, this API should be changed to take a ClassDesc instead.

liach replied:
In fact, the specification (javadoc) on this method is wrong: it will never take an internal name; it only takes descriptors and tries to simulate the result of `Class::getName`, since `ClassDesc` lacks such functionality.
